Elasticsearch addon

larmog opened the pull request:
Initial PR for collaboration.

luxas commented on addons/elasticsearch.yml:
```diff
@@ -0,0 +1,200 @@
+    component: elasticsearch
+    role: client
+spec:
+  type: LoadBalancer
```
Is this directly copied from somewhere? I don't think `LoadBalancer` exists on bare-metal

luxas commented on addons/elasticsearch.yml:
```diff
@@ -0,0 +1,200 @@
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: es-master
```
Why not in `kube-systemd`?

luxas commented on addons/elasticsearch.yml:
```diff
@@ -0,0 +1,200 @@
+          capabilities:
+            add:
+              - IPC_LOCK
+        image: larmog/elasticsearch-armhf:2.2.1
```
`kubernetesonarm/elasticsearch` and `imagePullPolicy: IfNotPresent`

luxas commented on addons/elasticsearch.yml:
```diff
@@ -0,0 +1,200 @@
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: "CLUSTER_NAME"
```
is the quoting required?

luxas commented on addons/elasticsearch.yml:
```diff
@@ -0,0 +1,200 @@
+          name: storage
+      volumes:
+          - emptyDir:
+              medium: ""
```
What does this line do?

larmog replied:
You mean `medium: ""`?

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+FROM larmog/armhf-alpine-java:jdk-8u73
+
+EXPOSE 9200 9300
```
Not required, makes it only slower :)

larmog replied:
You mean java? :)

luxas replied:
No `EXPOSE` adds an extra layer that's only required for `docker run -P`. The same for volume. I'm trying to make the layers as few as possible.

larmog replied:
The 9200 port is the default es port and 9300 is for node discovery in a es cluster. You need them.

luxas replied:
Yes, but they don't have to be defined in the Dockerfile for docker or Kubernetes to open the ports. Port 9200 and 9300 will be exposed without this layer.

larmog replied:
But it tells your users of your image that it expects these ports to be exposed. If you are looking for optimization then perhaps a better solution is to flatten you image after build.

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+
+# Install Elasticsearch.
+RUN apk add --update curl ca-certificates sudo && \
+  ( curl -Lskj https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/$VERSION/elasticsearch-$VERSION.tar.gz | \
```
`curl -sSL https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/$VERSION/elasticsearch-$VERSION.tar.gz | tar -xz --strip-components=1 -C es` makes the code much easier

larmog replied:
Yes agree, much of this comes from (@)pires. I've just rearranged stuff.

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+  apk del curl
+
+# Volume for Elasticsearch data
+VOLUME ["/data"]
```
not needed

larmog replied:
What you mean is not needed?

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+VOLUME ["/data"]
+
+# Copy configuration
+ADD config/logging.yml /elasticsearch/config
```
I think `COPY` is preferred over `ADD`

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+FROM larmog/armhf-alpine-java:jdk-8u73
```
I'm already working on a `luxas/java` image for this, will upload when I've got `libc` compiled for alpine.

larmog replied:
I know and we can change it when your done. But for now ...

luxas replied:
Yep, fine for now.

larmog commented:
Fixed, Ok?

luxas commented:
I'm on a journey until monday, will take a look after that

luxas commented on addons/elasticsearch.yaml:
```diff
@@ -0,0 +1,203 @@
+        - mountPath: /data
+          name: storage
+      volumes:
+          - emptyDir:
```
`emptyDir: {}` for consistency

luxas commented on addons/elasticsearch.yaml:
```diff
@@ -0,0 +1,203 @@
+          capabilities:
+            add:
+              - IPC_LOCK
+        image: kubernetesonarm/elasticsearch:2.2.1
```
Please `kubernetesonarm/elasticsearch` only and `imagePullPolicy: IfNotPresent` like other manifests

luxas commented on images/kubernetesonarm/elasticsearch/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+FROM larmog/armhf-alpine-java:jdk-8u73
+
+ENV VERSION 2.2.1
```
Please add `ELASTICSEARCH_VERSION` to `images/version.sh`, `COPY version.sh /`
Copy it to current dir via a `build.sh` file first.
Then `RUN . /version.sh && ... `
The `go` image is set up this way, you may find inspiration there (or I may do it as a followup)

luxas commented on images/kubernetesonarm/elasticsearch/run.sh:
```diff
@@ -0,0 +1,25 @@
+ulimit -l unlimited
+
+# run
+sudo -E -u elasticsearch /elasticsearch/bin/elasticsearch
```
I'm not an expert at this, but why is `sudo` required?
I see you're adding an es user, but why not use it directly as `root`?

luxas commented:
LGTM overall, some small things only :)

luxas commented:
LGTM

I'll fix the comments in a follow-up commit.

larmog commented:
Sorry, I wasn't finished :) I've been having some issues with my cluster and had to spend some time to fix it. I had some issues with `elasticsearch`, I couldn't get it stable. I haven't been able to get it to run stable for a longer period. Now I've added three more `odroid-c2` boards, so let's see if it makes any difference :). `Logstash-forwarder` wasn't a hit, it practically occupied the whole node. I've switched to `fluentd` and it looks more promising, but I can't say for sure.

luxas commented:
Okay, send follow up PRs when you're ready. It's on `dev`, so it's not such a big deal if it's unstable
